power_grid_model_c/power_grid_model/include/power_grid_model/main_core/update.hpp

```diff
@@ -130,4 +135,194 @@ inline void update_inverse(MainModelState<ComponentContainer> const& state, Forw
+                        return obj.id == first.id;
+                    });
+            });
+    }
+}
+
+template <class CompType>
+detail::UpdateCompProperties check_component_independence(ConstDataset const& update_data, Idx n_component) {
+    detail::UpdateCompProperties properties;
+    properties.name = CompType::name;
+    auto const component_idx = update_data.find_component(properties.name, false);
+    properties.is_columnar = update_data.is_columnar(properties.name);
+    properties.dense = update_data.is_dense(properties.name);
+    properties.uniform = update_data.is_uniform(properties.name);
+    properties.has_any_elements =
+        component_idx != invalid_index && update_data.get_component_info(component_idx).total_elements > 0;
+    properties.elements_ps_in_update =
+        properties.uniform ? update_data.uniform_elements_per_scenario(properties.name) : invalid_index;
+    properties.elements_in_base = n_component;
+
+    if (properties.is_columnar) {
+        detail::process_buffer_span<CompType>(
+            update_data.template get_columnar_buffer_span_all_scenarios<meta_data::update_getter_s, CompType>(),
+            properties);
+    } else {
+        detail::process_buffer_span<CompType>(
+            update_data.template get_buffer_span_all_scenarios<meta_data::update_getter_s, CompType>(), properties);
+    }
+
+    return properties;
+}
+
+inline void validate_update_data_independence(detail::UpdateCompProperties const& comp) {
+    if (comp.is_empty_component()) {
+        return; // empty dataset is still supported
+    }
+    auto const elements_ps = comp.get_n_elements();
+    assert(comp.uniform || elements_ps < 0);
+
+    if (elements_ps >= 0 && comp.elements_in_base < elements_ps) {
+        throw DatasetError("Update data has more elements per scenario than input data for component " + comp.name +
+                           "!");
+    }
+    if (comp.ids_part_na) {
+        throw DatasetError("Some IDs are not valid for component " + comp.name + " in update data!");
+    }
+    if (comp.ids_all_na && comp.elements_in_base != elements_ps) {
+        throw DatasetError("Update data without IDs for component " + comp.name +
+                           " has a different number of elements per scenario then input data!");
+    }
+}
+
+} // namespace detail
+
+// main core utils
+template <class CompType, class... ComponentTypes>
+constexpr auto index_of_component = main_core::utils::index_of_component<CompType, ComponentTypes...>;
+template <class... ComponentTypes> using SequenceIdx = main_core::utils::SequenceIdx<ComponentTypes...>;
+template <class... ComponentTypes> using ComponentFlags = main_core::utils::ComponentFlags<ComponentTypes...>;
+
+// get sequence idx map of a certain batch scenario
+template <typename CompType, class ComponentContainer>
+std::vector<Idx2D> get_component_sequence(MainModelState<ComponentContainer> const& state,
```

Comment:
We have many `get_component_sequence` defined at different places doing different things. Lets rename those too.

Reply:
The problematic one lives in `state_querries.hpp` and returns "sequence idx based on id or idx_2d", which is integer rather than a list of them. Other ones are doing the same thing. I would argue that the one in `state_querries.hpp` is named incorrectly. Let's fix that one in a separate PR maybe?

Reply:
@Jerry-Jinfeng-Guo Indeed!
Even without considering that problematic one, we have 2 variations here. 

I have a personal preference to avoid relying on function overloading unless for type variations or for special cases of templating etc.

Reply:
See [67929c5](https://github.com/PowerGridModel/power-grid-model/pull/828/commits/67929c5d2e9879f28f75620ce7de412ecaa0ce92). Let me know what you think.